### PR TITLE
Strapi

### PR DESCRIPTION
**Pull request by langdonholmes**
This PR bundles a few changes. It refactors how iTELL AI connects to databases and how tests are run.

- (working?) Strapi methods
- Adds pytest and organizes tests
- Updates Dockerfile and Makefile to run pytest as part of the build sequence. This should fail **before a deployment** if there is something majorly wrong. Currently, this just runs a few high-level "integration" tests. We should probably add more fine-grained "unit" tests at some point.
- (working?) retrieve_chunks functionality for chat
- Switches chat model to Orca2

TODO:

- Configure tests for no-GPU environment
- Move verbose Strapi query strings to the Strapi class
- More testing and bugfixes

### CODE REVIEW

File: models/base.py [hunk 1]
```diff
@@ -1,13 +1,33 @@
...
+    subsection_index: Optional[int] = Field(
+        0,
+        description="This field is deprecated and should only be used for legacy textbooks that are not in Strapi.",
+    )
+
+    @root_validator
+    def textbook_name_or_page_slug(cls, values):
+        if values.get("textbook_name") is None and values.get("page_slug") is None:
+            raise ValueError(
+                "Specify either textbook_name if the content is on SupaBase or page_slug if the conent is on Strapi."
```
Comment (jchoi92k) [inline models/base.py:28]:
```suggestion
                "Specify either textbook_name if the content is on SupaBase or page_slug if the content is on Strapi."

```

File: models/answer.py [hunk 2]
```diff
@@ -2,6 +2,8 @@
 
 
 class AnswerInput(InputBase):
+    page_slug: str
+    chunk_slug: str
     answer: str
```
Comment (jchoi92k) [inline models/answer.py:8]: I think it was nice we added error handling for `textbook_name` v. `page_slug`. Could do something similar here.

```suggestion

    @root_validator
    def answer_textbook_attributes_check(cls, values):
        textbook_name = values.get("textbook_name")
        chapter_index = values.get("chapter_index")
        section_index = values.get("section_index")
        subsection_index = values.get("subsection_index")
        if textbook_name is not None and (chapter_index is None or section_index is None or subsection_index is None):
            raise ValidationError("If providing textbook_name, provide values for chapter_index, section_index, and subsection_index.")

        return values
```

File: README.md [hunk 3]
```diff
@@ -1,31 +1,50 @@
...
+ **If no ENV is set, the default is production, which will likely fail on your system.**
 
-2. Run `python ./download_models.py` to download required models from SpaCy and Huggingface
-3. Make sure to create a `.env` file in the application root directory like the following. It should contain all the textbook names defined in the `src/models/textbooks.py`:
+1. Clone the repository and run `pip install -r requirements.txt` (use gpu-requirements.txt if you have a GPU)
+2. Run `python -m spacy download en_core_web_sm` to download required model from SpaCy
+3. Make sure to create a `.env` file in the application root directory like the following.
+   - If you are testing or working with legacy content databases, be sure to include the appropriate database credentials.
+   - If you are only working with Strapi-based deployments, you need only the Strapi credentials and the Vector-store credentials.
+   - Load the environment variables with `source .env` or by using the provided devcontainer.
```
Comment (jchoi92k) [inline README.md:35]: Might make sense to create an `.env.sample` file and add some comments about legacy v. new system there. I can make one if you'd like if you send me your most up-to-date env file.

File: tests/test_summary_eval_supabase.py [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+    response = client.post(
+        "/score/summary",
+        json={
+            "textbook_name": "macroeconomics-2e",
+            "chapter_index": "1",
+            "section_index": "1",
+            "summary": """Think about it this way: In 2015 the labor force in the United States contained over 158 million workers, according to the U.S. Bureau of Labor Statistics. The total land area was 3,794,101 square miles. While these are certainly large numbers, they are not infinite. Because these resources are limited, so are the numbers of goods and services we produce with them. Combine this with the fact that human wants seem to be virtually infinite, and you can see why scarcity is a problem.',""",  # noqa: E501
+        },
+    )
+    assert response.json()["containment"] > 0.9
```
Comment (jchoi92k) [inline tests/test_summary_eval_supabase.py:11]: Unless we're specifically testing containment?

```suggestion
    assert response.status_code == 200
```

File: src/chat.py [hunk 5]
```diff
@@ -32,21 +48,28 @@ async def moderated_chat(chat_input: ChatInput) -> AsyncGenerator[bytes, None]:
...
+            text=chat_input.message,
+            match_count=1,
+        )
+    )
     if relevant_chunks:
         additional_context += "\n# This is some additional context:"
-        for chunk in relevant_chunks:
-            additional_context += f"\n{chunk['clean_text']}"
+        for chunk in relevant_chunks.matches:
+            additional_context += f"\n{chunk.content}"
```
Comment (jchoi92k) [inline src/chat.py:72]: We have some pretty long chunks that can get to 1000+ words. Can't find information on Orca 2's maximum input length, but with the way things are set up ```prompt = "".join([preface, sample_conversation, additional_context, history, msg])```, user's message might become truncated. 

We could create a threshold chunk length before adding them to `additional_context`.

**Comment by jchoi92k** [comment]
Looks good to me overall! Props to both @langdonholmes and @vaiibhavgupta. Left a few comments to flag some potential, minor issues. Let me know if anything needs clarification.

File: src/chat.py (hunk 5 above)
Comment (langdonholmes) [reply src/chat.py:72]: Yes, this is still a work in progress. I wish we had decided to set stricter chunk size limits in Strapi, but we can work around it. I also do not think we are 100% on the format for the chat history.

Anyway, Orca 2 has a sequence limit of 4096 in its HuggingFace config (not the most reliable source, but there ya go). For now, I'll just capture the first 2,500 characters (roughly 600 tokens) of each chunk. This should be the entire chunk in most cases.

File: tests/test_summary_eval_supabase.py (hunk 4 above)
Comment (langdonholmes) [reply tests/test_summary_eval_supabase.py:11]: This was copied from an old unit test. I agree that it makes more sense to assert the status code here.

File: models/answer.py [hunk 6]
```diff
@@ -2,6 +2,8 @@
 
 
 class AnswerInput(InputBase):
+    page_slug: str
+    chunk_slug: str
```
[the remark this answers is not in the document]
Comment (langdonholmes) [reply models/answer.py:6]: Good catch. My tests should've errored on this, so I'll need to revise those accordingly as well.

**Comment by langdonholmes** [comment]
There is still some outstanding work, but I am going to merge this in since we need to get some of these features up and running quickly.

**Pull request merged**